primitives/utils/src/mpsc.rs

```diff
@@ -0,0 +1,220 @@
+        UnboundedReceiver, UnboundedSender,
+        TryRecvError, TrySendError, SendError
+    };
+    use futures::{sink::Sink, task::{Poll, Context}, stream::Stream};
+    use std::pin::Pin;
+    use crate::metrics::UNBOUNDED_CHANNELS_COUNTER;
+
+    /// Wrapper Type around `UnboundedSender` that increases the global
+    /// measure when a message is added
+    #[derive(Debug)]
+    pub struct TracingUnboundedSender<T>(&'static str, UnboundedSender<T>);
+
+    /// Wrapper Type around `UnboundedReceiver` that decreases the global
+    /// measure when a message is polled
+    #[derive(Debug)]
+    pub struct TracingUnboundedReceiver<T>(&'static str, UnboundedReceiver<T>);
+
+    /// Wrapper around `mpsc::unbounded` that tracks the in- and outflow via
+    /// `UNBOUNDED_CHANNELS_COUNTER`
+    pub fn tracing_unbounded<T>(key: &'static str) ->(TracingUnboundedSender<T>, TracingUnboundedReceiver<T>) {
+        let (s, r) = mpsc::unbounded();
+        (TracingUnboundedSender(key.clone(), s), TracingUnboundedReceiver(key,r))
+    }
+
+    impl<T> TracingUnboundedSender<T> {
+        /// Proxy function to mpsc::UnboundedSender
+        pub fn poll_ready(&self, ctx: &mut Context) -> Poll<Result<(), SendError>> {
+            self.1.poll_ready(ctx)
+        }
+
+        /// Proxy function to mpsc::UnboundedSender
+        pub fn is_closed(&self) -> bool {
+            self.1.is_closed()
+        }
+
+        /// Proxy function to mpsc::UnboundedSender
+        pub fn close_channel(&self) {
+            self.1.close_channel()
+        }
+
+        /// Proxy function to mpsc::UnboundedSender
+        pub fn disconnect(&mut self) {
+            self.1.disconnect()
+        }
+
+        /// Proxy function to mpsc::UnboundedSender
+        pub fn start_send(&mut self, msg: T) -> Result<(), SendError> {
+            self.1.start_send(msg)
+        }
+
+        /// Proxy function to mpsc::UnboundedSender
+        pub fn unbounded_send(&self, msg: T) -> Result<(), TrySendError<T>> {
+            self.1.unbounded_send(msg).map(|s|{
+                UNBOUNDED_CHANNELS_COUNTER.with_label_values(self.0).incr();
+                s
+            })
+        }
+
+        /// Proxy function to mpsc::UnboundedSender
+        pub fn same_receiver(&self, other: &UnboundedSender<T>) -> bool {
+            self.1.same_receiver(other)
+        }
+    }
+
+    impl<T> Clone for TracingUnboundedSender<T> {
+        fn clone(&self) -> Self {
+            Self(self.0.clone(), self.1.clone())
+        }
+    }
+
+    impl<T> TracingUnboundedReceiver<T> {
+
+        fn consume(&mut self) {
+            // consume all items, make sure to reflect the updated count
+            let mut count = 0;
+            while let Ok(Some(..)) = self.try_next() {
+                count += 1;
+            }
+
+            // and discount the messages
+            if count > 0 {
+                UNBOUNDED_CHANNELS_COUNTER.with_label_values(self.0).decr_by(count);
```

Comment:
WIth the suggestion above of two counters, this would then increase the `received` counter.